File: ecommerce/backend/app/schemas/loyalty.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.common import TimestampMixin


class LoyaltyAccountResponse(TimestampMixin):
    id: int
    customer_id: int
    points_earned: int = 0
    points_redeemed: int = 0
    points_expired: int = 0
    current_balance: int = 0

    model_config = ConfigDict(from_attributes=True)


class LoyaltyTransactionResponse(BaseModel):
    id: int
    account_id: int
    transaction_type: str
    points: int
    reference_type: str | None = None
    reference_id: int | None = None
    description: str | None = None
    expiry_date: date | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LoyaltyTransactionListResponse(BaseModel):
    transactions: list[LoyaltyTransactionResponse]
    total: int


class LoyaltySettings(BaseModel):
    points_per_currency: float = 1.0
    redemption_rate: float = 0.05
    min_redemption_points: int = 100
    points_expiry_days: int = 365


class LoyaltyRedeemRequest(BaseModel):
    points: int


class LoyaltyRedeemResponse(BaseModel):
    points_redeemed: int
    amount_off: float
    remaining_balance: int

File: mobile/shared_core/lib/src/models/loyalty.dart
class LoyaltyAccount {
  final int id;
  final int customerId;
  final int pointsEarned;
  final int pointsRedeemed;
  final int pointsExpired;
  final int currentBalance;

  LoyaltyAccount({
    required this.id,
    required this.customerId,
    this.pointsEarned = 0,
    this.pointsRedeemed = 0,
    this.pointsExpired = 0,
    this.currentBalance = 0,
  });

  factory LoyaltyAccount.fromJson(Map<String, dynamic> json) {
    return LoyaltyAccount(
      id: json['id'],
      customerId: json['customer_id'],
      pointsEarned: json['points_earned'] ?? 0,
      pointsRedeemed: json['points_redeemed'] ?? 0,
      pointsExpired: json['points_expired'] ?? 0,
      currentBalance: json['current_balance'] ?? 0,
    );
  }

  double get approximateValue => currentBalance * 0.05;
}

class LoyaltyTransaction {
  final int id;
  final String transactionType;
  final int points;
  final String? referenceType;
  final int? referenceId;
  final String? description;
  final String? expiryDate;
  final String createdAt;

  LoyaltyTransaction({
    required this.id,
    required this.transactionType,
    required this.points,
    this.referenceType,
    this.referenceId,
    this.description,
    this.expiryDate,
    required this.createdAt,
  });

  factory LoyaltyTransaction.fromJson(Map<String, dynamic> json) {
    return LoyaltyTransaction(
      id: json['id'],
      transactionType: json['transaction_type'] ?? '',
      points: json['points'] ?? 0,
      referenceType: json['reference_type'],
      referenceId: json['reference_id'],
      description: json['description'],
      expiryDate: json['expiry_date'],
      createdAt: json['created_at'] ?? '',
    );
  }

  bool get isEarned => transactionType == 'earn';
  bool get isRedeemed => transactionType == 'redeem';
  bool get isExpired => transactionType == 'expire';
}
